### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by glelouet**
should be done manually as many modifications are whitespace / import / brackets automatically performed by my IDE.

### CODE REVIEW

File: src/main/java/org/chocosolver/solver/search/IResolutionHelper.java [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+	 *          optional criterions to stop the search before finding all/best solution
+	 * @return
+	 *         <ul>
+	 *         <li><tt>null</tt> if the problem has no solution or a stop criterion stops the search before finding a
+	 *         first solution</li>
+	 *         <li>a {@link Solution} if at least one solution has been found. The solution is proven to be optimal if no
+	 *         stop criterion stops the search.</li>
+	 *         </ul>
+	 */
+	default Solution findOptimalSolution(IntVar objective, boolean maximize, Criterion... stop) {
```
Comment (cprudhom) [inline src/main/java/org/chocosolver/solver/search/IResolutionHelper.java]: Codacy Issue found: [The method findOptimalSolution() has an NPath complexity of 200](https://www.codacy.com/app/cprudhom/choco-solver/file/2464610826/issues/source?bid=3218989&fileBranchId=3256349#l232)

File: src/main/java/org/chocosolver/solver/search/measure/MeasuresRecorder.java [hunk 2]
```diff
@@ -40,326 +42,348 @@
...
+				f.setAccessible(true);
+			}
+			if (f.getType().isPrimitive()) {
+				try {
+					f.set(ret, f.get(this));
+				} catch (IllegalArgumentException | IllegalAccessException e) {
+					// XXX
+					e.printStackTrace();
+				}
+			} else {
```
Comment (cprudhom) [inline src/main/java/org/chocosolver/solver/search/measure/MeasuresRecorder.java]: Codacy Issue found: [Avoid empty if statements](https://www.codacy.com/app/cprudhom/choco-solver/file/2464610881/issues/source?bid=3218989&fileBranchId=3256349#l338)

**Comment by cprudhom** [comment]
I'll merge it asap.
I think we can go a step forward and try to avoid any reference to the `Model` in `MeasuresRecorder`.
That means some new methods should be added, like:
```java
void setBestSolutionValue(Number n);
void setStatus(Status s); // among: RUNNING, KILLED, STOP, COMPLETE
void setModelName(String s); // needed for toString() 
```

Plus (again), we should consider adding `IFeatures` and `IFeaturesRecorder` related to model statistics (nb vars, cstrs, etc) and then removed the number of variables and constraints from `toString()` of `MeasuresRecorder`.

@jgFages any remarks on that point ?
@arnaud-m would this fits (partially?) your need

**Comment by jgFages** [comment]
ok

**Comment by arnaud-m** [comment]
Yes, it looks good. I will try to update my code to the develop branch next week.
Then, I will give you more feedback.

**Comment by cprudhom** [comment]
Euh.. I didn't noticed that since `findAllSolutions()` returns a `Stream` it needs to be evaluated (through `count()` for instance, otherwise, nothing happen...
Maybe we should consider duplicating the methods.

**Pull request merged**